File: src/automation/ActionValidator.ts
import { Action, ActionType, ValidationResult } from './types';

export class ActionValidator {
  private static readonly VALID_ACTION_TYPES = new Set<ActionType>(Object.values(ActionType));
  private static readonly MIN_TIMEOUT_MS = 100;
  private static readonly MAX_TIMEOUT_MS = 60000;
  private static readonly MAX_RETRY_COUNT = 5;

  private static readonly RESTRICTED_PACKAGES = new Set([
    'com.google.android.apps.nfc',
    'com.android.settings.security',
    'com.google.android.apps.walletnfcrel',
    'com.phonepe.app',
    'net.one97.paytm',
    'com.google.android.apps.authenticator2',
  ]);

  /**
   * Validates an AI-generated Action object before execution.
   * Guarantees that malformed output will never crash the system.
   */
  public static validate(rawAction: Partial<Action>, sensitiveAppsDenylist: string[] = []): ValidationResult {
    const errors: string[] = [];

    if (!rawAction) {
      return { valid: false, errors: ['Action object is null or undefined'] };
    }

    // 1. Validate Action ID
    const actionId = typeof rawAction.actionId === 'string' && rawAction.actionId.trim().length > 0
      ? rawAction.actionId.trim()
      : `act-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`;

    // 2. Validate Action Type
    if (!rawAction.actionType || !this.VALID_ACTION_TYPES.has(rawAction.actionType)) {
      errors.push(`Invalid actionType: "${rawAction.actionType}". Supported: ${Array.from(this.VALID_ACTION_TYPES).join(', ')}`);
    }

    // 3. Validate Parameters
    const parameters = typeof rawAction.parameters === 'object' && rawAction.parameters !== null
      ? { ...rawAction.parameters }
      : {};

    // 4. Validate Specific Action Type Requirements
    if (rawAction.actionType === ActionType.APP_ACTION) {
      const appName = parameters.appName || parameters.packageName;
      if (!appName || typeof appName !== 'string') {
        errors.push('APP_ACTION requires "appName" or "packageName" in parameters');
      } else {
        const lowerApp = appName.toLowerCase();
        if (
          this.RESTRICTED_PACKAGES.has(lowerApp) ||
          sensitiveAppsDenylist.some((s) => lowerApp.includes(s.toLowerCase()))
        ) {
          errors.push(`Target application "${appName}" is restricted by Android security policy`);
        }
      }
    } else if (rawAction.actionType === ActionType.UI_ACTION) {
      const operation = parameters.operation;
      const validOps = ['click', 'long_click', 'focus', 'select', 'scroll', 'back', 'find', 'wait'];
      if (!operation || !validOps.includes(operation)) {
        errors.push(`UI_ACTION requires valid "operation" (${validOps.join(', ')}), got: ${operation}`);
      }
    } else if (rawAction.actionType === ActionType.GESTURE_ACTION) {
      const gesture = parameters.gesture;
      const validGestures = ['tap', 'long_press', 'double_tap', 'swipe_up', 'swipe_down', 'swipe_left', 'swipe_right', 'drag', 'scroll'];
      if (!gesture || !validGestures.includes(gesture)) {
        errors.push(`GESTURE_ACTION requires valid "gesture" (${validGestures.join(', ')}), got: ${gesture}`);
      }
    } else if (rawAction.actionType === ActionType.TEXT_ACTION) {
      const operation = parameters.operation;
      const validTextOps = ['enter', 'replace', 'append', 'clear', 'submit', 'focus'];
      if (!operation || !validTextOps.includes(operation)) {
        errors.push(`TEXT_ACTION requires valid "operation" (${validTextOps.join(', ')}), got: ${operation}`);
      }
      if (['enter', 'replace', 'append'].includes(operation) && typeof parameters.text !== 'string') {
        errors.push(`TEXT_ACTION with operation "${operation}" requires a "text" string parameter`);
      }
    } else if (rawAction.actionType === ActionType.DELAY_ACTION) {
      const ms = Number(parameters.durationMs);
      if (isNaN(ms) || ms < 0 || ms > 30000) {
        errors.push('DELAY_ACTION requires "durationMs" between 0 and 30000ms');
      }
    }

    // 5. Validate Timeout
    let timeout = Number(rawAction.timeout);
    if (isNaN(timeout) || timeout < this.MIN_TIMEOUT_MS) {
      timeout = 10000; // Default 10s
    } else if (timeout > this.MAX_TIMEOUT_MS) {
      timeout = this.MAX_TIMEOUT_MS;
    }

    // 6. Validate Retry Count
    let retryCount = Number(rawAction.retryCount);
    if (isNaN(retryCount) || retryCount < 0) {
      retryCount = 2; // Default 2 retries
    } else if (retryCount > this.MAX_RETRY_COUNT) {
      retryCount = this.MAX_RETRY_COUNT;
    }

    // 7. Validate Permissions
    const requiredPermissions = Array.isArray(rawAction.requiredPermissions)
      ? rawAction.requiredPermissions.filter((p) => typeof p === 'string')
      : ['android.permission.BIND_ACCESSIBILITY_SERVICE'];

    // Sanitized action result
    const sanitizedAction: Action = {
      actionId,
      actionType: rawAction.actionType as ActionType,
      title: rawAction.title || `${rawAction.actionType} execution`,
      parameters,
      timeout,
      retryCount,
      retriesAttempted: 0,
      requiredPermissions,
      executionState: rawAction.executionState || (errors.length === 0 ? ('VALIDATING' as any) : ('FAILED' as any)),
      priority: typeof rawAction.priority === 'number' ? rawAction.priority : 1,
    };

    return {
      valid: errors.length === 0,
      errors,
      sanitizedAction: errors.length === 0 ? sanitizedAction : undefined,
    };
  }
}

File: android/app/src/main/java/com/nova/assistant/automation/ActionValidator.kt
package com.nova.assistant.automation

import java.util.UUID

/**
 * Production Action Validator
 * Safeguards system from malformed AI output and security violations.
 */
object ActionValidator {

    private const val MIN_TIMEOUT_MS = 100L
    private const val MAX_TIMEOUT_MS = 60000L
    private const val MAX_RETRY_COUNT = 5

    private val RESTRICTED_PACKAGES = setOf(
        "com.google.android.apps.nfc",
        "com.android.settings.security",
        "com.google.android.apps.walletnfcrel",
        "com.phonepe.app",
        "net.one97.paytm",
        "com.google.android.apps.authenticator2"
    )

    fun validate(rawAction: Action, sensitiveDenylist: List<String> = emptyList()): ValidationResult {
        val errors = mutableListOf<String>()

        // 1. Validate ID
        val actionId = if (rawAction.actionId.isNotBlank()) {
            rawAction.actionId
        } else {
            "act-${UUID.randomUUID()}"
        }

        // 2. Validate Parameters based on ActionType
        val params = rawAction.parameters
        when (rawAction.actionType) {
            ActionType.APP_ACTION -> {
                val appOrPkg = params["appName"] as? String ?: params["packageName"] as? String
                if (appOrPkg.isNullOrBlank()) {
                    errors.add("APP_ACTION requires 'appName' or 'packageName'")
                } else {
                    val lower = appOrPkg.lowercase()
                    if (RESTRICTED_PACKAGES.contains(lower) || sensitiveDenylist.any { lower.contains(it.lowercase()) }) {
                        errors.add("Package '$appOrPkg' is restricted by security policy")
                    }
                }
            }
            ActionType.UI_ACTION -> {
                val op = params["operation"] as? String
                val validOps = setOf("click", "long_click", "focus", "select", "scroll", "back", "find", "wait")
                if (op == null || !validOps.contains(op)) {
                    errors.add("UI_ACTION requires valid 'operation' (click, long_click, focus, scroll, back, find, wait)")
                }
            }
            ActionType.GESTURE_ACTION -> {
                val gesture = params["gesture"] as? String
                val validGestures = setOf("tap", "long_press", "double_tap", "swipe_up", "swipe_down", "swipe_left", "swipe_right", "drag", "scroll")
                if (gesture == null || !validGestures.contains(gesture)) {
                    errors.add("GESTURE_ACTION requires valid 'gesture' name")
                }
            }
            ActionType.TEXT_ACTION -> {
                val op = params["operation"] as? String
                val validOps = setOf("enter", "replace", "append", "clear", "submit", "focus")
                if (op == null || !validOps.contains(op)) {
                    errors.add("TEXT_ACTION requires valid 'operation' (enter, replace, append, clear, submit, focus)")
                }
                if (setOf("enter", "replace", "append").contains(op) && params["text"] !is String) {
                    errors.add("TEXT_ACTION operation '$op' requires a string 'text' parameter")
                }
            }
            ActionType.DELAY_ACTION -> {
                val duration = (params["durationMs"] as? Number)?.toLong()
                if (duration == null || duration < 0 || duration > 30000L) {
                    errors.add("DELAY_ACTION requires 'durationMs' between 0 and 30000")
                }
            }
            else -> {
                // Other action types permitted with default safety limits
            }
        }

        // 3. Validate Timeout
        val sanitizedTimeout = rawAction.timeout.coerceIn(MIN_TIMEOUT_MS, MAX_TIMEOUT_MS)

        // 4. Validate Retry Count
        val sanitizedRetry = rawAction.retryCount.coerceIn(0, MAX_RETRY_COUNT)

        val sanitized = rawAction.copy(
            actionId = actionId,
            timeout = sanitizedTimeout,
            retryCount = sanitizedRetry
        )

        return ValidationResult(
            valid = errors.isEmpty(),
            errors = errors,
            sanitizedAction = if (errors.isEmpty()) sanitized else null
        )
    }
}
